## Introduction

A typical data science pipeline involves data acquisition, extraction (web scraping in HW1), cleaning (time series data in HW2), integration (data blocking then matching), and final analysis. The goal of this tutorial is to introduce you a useful library, py_entitymatching, for data blocking in data integration stage.

What is __blocking__? This stage comes after data cleaning and before data matching. We are familiar with processing single source data but if we have more than one source of data, we need to compare them for further analysis. However, you cannot compare an apple to an orange, so we need to match the identity first. For example, look at the left and right tables below. Based on the name similarity and location, we pair Dave Smith on the left to the David Smith on the right, and pair Dan Smith on the left to the Daniel B. Smith on the right. There exist 2*3=6 possible pairs (each people on the left tries to match each people on the right) and we try to block the other 4 pairs that not appear to match. 

Name         |          City | States        | |  Name         |          City | States
------------ | ------------- | ------------  |-|  ------------ | ------------- | ------------
Joe Brown    | Pittsburgh    | PA            | |  David Smith  | Los Angeles   | CA
Dave Smith   | Los Angeles   | CA            | |  Daniel B. Smith | Sacramento | CA
Dan Smith    | Sacramento    | CA   

As the size of the two tables grows, it is hard to manually find all matches (2 tables with 1,000 rows each will generate 1,000,000 candidate pairs). The __goal__ of data blocking is to come up with blockers that reduce the size of candidate pairs. After data blocking and data matching, the next step depends on your goal, you can specifically analyze the matched pairs or you can merge the tables with deduplication so you have a larger data pool.

We will cover the following topics in this tutorial:
- [Install the library](#Install-the-library)
- [Pre-processing](#Pre-processing)
- [Blockers](#Blockers)
- [What's Next](#What's-Next)
- [Summary and references](#Summary-and-references)



## Install the library

Make sure you have Anaconda installed, then run the following commands:

    $ conda install -c uwmagellan py_entitymatching

    $ conda uninstall py_entitymatching

    $ pip install 'py_entitymatching==0.2.0'

The first command installs all the dependencies of the library, second and third commands substitute the library to the stable version. I use the older version of the library because I found the newest version has compatibility issue on some machines.

Now make sure the import works without problem.

In [1]:
import py_entitymatching as em
import os
import pandas as pd
import numpy as np

In this tutorial, I'll demonstrate how to build different blockers provided by py_entitymatching library and how to use them on my data set. I collected roughly 8000 books information from the web (search using keyword Java and data etc.), with 5000 transformed from Barnes&Noble’s (tableA.csv) and 3000 transformed from Goodreads (tableB.csv). Their attributes are transformed into the same format:

**id, title, authors, ISBN13, pages, publisher, publishedYear, publishedMonth, publishedDay**

I only extracted **basic information** of the books from the web. To conduct meaning analysis in the real world scenario, you can include attributes such as book price, reviews, book rating when you do web scraping. I will use the word 'attribute' to represent the column names.


In [2]:
# filepaths for table A and B. 
cur_path = os.getcwd()
path_A = os.path.join(cur_path, 'tableA.csv')
path_B = os.path.join(cur_path, 'tableB.csv')

In [3]:
# read table A; 'id' as the key attribute
A = em.read_csv_metadata(path_A, key='id')
# read table B; 'id' as the key attribute
B = em.read_csv_metadata(path_B, key='id')

Metadata file is not present in the given path; proceeding to read the csv file.
Metadata file is not present in the given path; proceeding to read the csv file.


The above code reads tableA and tableB. The entity matching library is built upon pandas. Make sure you have a column as primary key (unique value for each row). In my case, I set 'id' as key. You are free to use pandas operations to the table you've loaded.

Take a look at the data sets. Scroll down to have a rough understanding of tableA and tableB. Do you see any problem?

In [4]:
A.shape

(5279, 9)

In [5]:
A.head(15)

id  \
0    0   
1    1   
2    2   
3    3   
4    4   
5    5   
6    6   
7    7   
8    8   
9    9   
10  10   
11  11   
12  12   
13  13   
14  14   

                                                                                                  title  \
0                                           The Java Tutorial: A Short Course on the Basics / Edition 6   
1                                                               Java : Complete Java Programming Guide.   
2                                   Java Programs to Accompany Programming Logic and Design / Edition 8   
3                                      Raspberry Pi with Java: Programming the Internet of Things (IoT)   
4                                  Java Generics and Collections: Speed Up the Java Development Process   
5                                                      Iron-Clad Java: Building Secure Web Applications   
6                                                        Java SE 8 for the Really Impatient / Edition 1   
7                         Programming Groovy 2: Dynamic Productivity for the Java Developer / Edition 1   
8   Mikrographie des Holzes der auf Java vorkommenden Baumarten, im Auftrage des Kolonial-Ministeriu...   
9                                                                                Using Pointers in Java   
10                                             Algorithms in Java, Part 5: Graph Algorithms / Edition 3   
11                                    Nelles Guide to Indonesia - Sumatra, Java, Bali, Lombok, Sulawesi   
12                                                   Think Java: How to Think Like a Computer Scientist   
13                                     Mostly Java--A Dictionary Of Distributed Computing (5th Edition)   
14                     JavaFX Script: Dynamic Java Scripting for Rich Internet/Client-side Applications   

                                                                          authors  \
0   Raymond Gallardo,Scott Hommel,Sowmya Kannan,Joni Gordon,Sharon Biocca Zakhour   
1                                                            Harry. H. Chaudhary.   
2                                                                    Jo Ann Smith   
3                                                       Stephen Chin,James Weaver   
4                                                  Maurice Naftalin,Philip Wadler   
5                                                     Jim Manico,August Detlefsen   
6                                                                Cay S. Horstmann   
7                                                              Venkat Subramaniam   
8                                                        Hindrik Haijo Janssonius   
9                                                                Aatif Ahmad Khan   
10                      Robert Sedgewick,Michael Schidlowsky,Michael Schindlowsky   
11                                                  Seven Hills Book Distributors   
12                                                                Downey,Mayfield   
13                                                                     John Hines   
14                                                                   James Weaver   

           ISBN13  pages                            publisher publishedYear  \
0   9780134034089  864.0                       Addison-Wesley          2014   
1   2940046303308    0.0                  Harry & Associates.          2014   
2   9781285867403  200.0                     Cengage Learning          2014   
3   9780071842013  336.0  McGraw-Hill Professional Publishing          2015   
4   9780596527754  286.0         O'Reilly Media, Incorporated          2006   
5   9780071835886  304.0  McGraw-Hill Professional Publishing          2014   
6   9780321927767  240.0                       Addison-Wesley          2014   
7   9781937785307  370.0      Pragmatic Programmers, LLC, The          2013   
8   9781314978940  626.0                 HardPress Publish

In [6]:
B.shape

(3785, 9)

In [7]:
B.head(15)

id  \
0    0   
1    1   
2    2   
3    3   
4    4   
5    5   
6    6   
7    7   
8    8   
9    9   
10  10   
11  11   
12  12   
13  13   
14  14   

                                                                                      title  \
0                                         The Java Language Specification (The Java Series)   
1                                                                   Java Performance Tuning   
2                 Java Software Solutions (Java 5.0 version): Foundations of Program Design   
3                                                Data Structures and Abstractions with Java   
4                                                                             Java and XSLT   
5                                  Java: An Introduction to Problem Solving and Programming   
6                                         Core Java 2, Volume I--Fundamentals (Core Series)   
7                                                                    Big Java: Late Objects   
8                                                                  Just Java 2 [With CDROM]   
9                               Enterprise JavaBeans (Java Series (O'Reilly & Associates).)   
10                                                    Learn Objective-C for Java Developers   
11  Java Server Programming For Professionals, 2nd Edition   Covers Java Ee 5 (Book/Cd Rom)   
12                                                                      Programming in Java   
13                                                                  Java Testing with Spock   
14                                        Sun Performance and Tuning: Java and the Internet   

                                                     authors        ISBN13  \
0      James Gosling,Bill Joy,Guy L. Steele Jr.,Gilad Bracha  9.780321e+12   
1                                               Jack Shirazi  9.780596e+12   
2                                John   Lewis,William Loftus  9.780321e+12   
3                         Frank M. Carrano,Walter J. Savitch  9.780132e+12   
4                                              Eric M. Burke  9.780596e+12   
5                                          Walter J. Savitch  9.780131e+12   
6                              Cay S. Horstmann,Gary Cornell  9.780131e+12   
7                                           Cay S. Horstmann  9.781118e+12   
8                                       Peter van der Linden  9.780130e+12   
9                                      Richard Monson-Haefel  9.781566e+12   
10                                             James Bucanek  9.781430e+12   
11  Ivan Bayross,Sharanam Shah,Cynthia Bayross,Vaishali Shah  9.788184e+12   
12                         Sachin Malhotra,Saurabh Choudhary  9.780198e+12   
13                                    Konstantinos Kapelonis  9.781617e+12   
14                   Adrian Cockcroft,Sun Microsystems Press  9.780131e+12   

     pages                          publisher  publishedYear  publishedMonth  \
0    651.0        Addison-Wesley Professional         2005.0             6.0   
1    600.0                     O'Reilly Media         2003.0             1.0   
2    944.0  Addison Wesley Publishing Company         2005.0             7.0   
3    998.0                      Prentice Hall         2006.0             8.0   
4    528.0                     O'Reilly Media         2001.0             9.0   
5   1060.0                      Prentice Hall         2004.0            12.0   
6    784.0                      Prentice Hall         2004.0             8.0   
7   1056.0                              Wiley         2012.0             2.0   
8   1098.0                  Prentice Hall PTR         2001.0            12.0   
9    489.0                           O'Reilly         2001.0             9.0   
10   520.0                             Apress         2009.0             9.0   
11     NaN                                NaN            NaN             NaN   
12   656.0       Oxford Universit

Here are some obvious ones:
1. Type mismatch (ISBN13 of tableA and tableB are interpreted differently, published dates are int/float)
2. Missing values (0 or NaN)

We need some data pre-processing before we use the library.

## Pre-processing

Although blocking is the focus, I think it is necessary to touch a bit on data cleaning. ALWAYS remember to clean the data and convert to the right data types! Otherwise, our library will not produce what you desired. I have situations where after 10 minutes running blockers, the output table is none because the data type of the compared attributes are different. 

When I apply the to_numeric operation, I set the errors to 'coerce' so invalid conversions will be set as NaN. Then I replace value 0 to NaN because ISBN, page and published date should not have 0. Lastly, I can drop rows that have NaN values in it.

In [8]:
A[['id','ISBN13','pages','publishedYear','publishedMonth','publishedDay']] = \
    A[['id','ISBN13','pages','publishedYear','publishedMonth','publishedDay']].apply(pd.to_numeric,errors='coerce')
B[['id','ISBN13','pages','publishedYear','publishedMonth','publishedDay']] = \
    B[['id','ISBN13','pages','publishedYear','publishedMonth','publishedDay']].apply(pd.to_numeric,errors='coerce')
A[['ISBN13','pages','publishedYear','publishedMonth','publishedDay']] = \
    A[['ISBN13','pages','publishedYear','publishedMonth','publishedDay']].replace(0,np.nan)
B[['ISBN13','pages','publishedYear','publishedMonth','publishedDay']] = \
    B[['ISBN13','pages','publishedYear','publishedMonth','publishedDay']].replace(0,np.nan)
A.dropna(axis=0, how='any', inplace=True)
B.dropna(axis=0, how='any', inplace=True)
A[['id','ISBN13','pages','publishedYear','publishedMonth','publishedDay']] = \
    A[['id','ISBN13','pages','publishedYear','publishedMonth','publishedDay']].astype(np.int64)
B[['id','ISBN13','pages','publishedYear','publishedMonth','publishedDay']] = \
    B[['id','ISBN13','pages','publishedYear','publishedMonth','publishedDay']].astype(np.int64)

Look at the tables after cleaning below. You can see we dropped roughly 500 rows from tableA and more than 1000 rows from tableB. I'm dropping all rows with missing values, but you can decide your own rules on how to deal with missing values as long as it doesn't mislead your goal (and you have more data samples if you relax your rules).

We have roughly 12,000,000 candidate pairs before we enter the blocking stage (4806*2593). Let's see how each blocker reduces such huge candidate size. 

In [9]:
A.shape

(4806, 9)

In [10]:
A.head()

id                                                                 title  \
0   0           The Java Tutorial: A Short Course on the Basics / Edition 6   
2   2   Java Programs to Accompany Programming Logic and Design / Edition 8   
3   3      Raspberry Pi with Java: Programming the Internet of Things (IoT)   
4   4  Java Generics and Collections: Speed Up the Java Development Process   
5   5                      Iron-Clad Java: Building Secure Web Applications   

                                                                         authors  \
0  Raymond Gallardo,Scott Hommel,Sowmya Kannan,Joni Gordon,Sharon Biocca Zakhour   
2                                                                   Jo Ann Smith   
3                                                      Stephen Chin,James Weaver   
4                                                 Maurice Naftalin,Philip Wadler   
5                                                    Jim Manico,August Detlefsen   

          ISBN13  pages                            publisher  publishedYear  \
0  9780134034089    864                       Addison-Wesley           2014   
2  9781285867403    200                     Cengage Learning           2014   
3  9780071842013    336  McGraw-Hill Professional Publishing           2015   
4  9780596527754    286         O'Reilly Media, Incorporated           2006   
5  9780071835886    304  McGraw-Hill Professional Publishing           2014   

   publishedMonth  publishedDay  
0              12            26  
2               3             3  
3              11             3  
4              10             1  
5               9             9

In [11]:
A.dtypes

id                 int64
title             object
authors           object
ISBN13             int64
pages              int64
publisher         object
publishedYear      int64
publishedMonth     int64
publishedDay       int64
dtype: object

In [12]:
B.shape

(2593, 9)

In [13]:
B.head()

id  \
0   0   
1   1   
2   2   
3   3   
4   4   

                                                                       title  \
0                          The Java Language Specification (The Java Series)   
1                                                    Java Performance Tuning   
2  Java Software Solutions (Java 5.0 version): Foundations of Program Design   
3                                 Data Structures and Abstractions with Java   
4                                                              Java and XSLT   

                                                 authors         ISBN13  \
0  James Gosling,Bill Joy,Guy L. Steele Jr.,Gilad Bracha  9780321246783   
1                                           Jack Shirazi  9780596003777   
2                            John   Lewis,William Loftus  9780321322036   
3                     Frank M. Carrano,Walter J. Savitch  9780132370455   
4                                          Eric M. Burke  9780596001438   

   pages                          publisher  publishedYear  publishedMonth  \
0    651        Addison-Wesley Professional           2005               6   
1    600                     O'Reilly Media           2003               1   
2    944  Addison Wesley Publishing Company           2005               7   
3    998                      Prentice Hall           2006               8   
4    528                     O'Reilly Media           2001               9   

   publishedDay  
0             1  
1            28  
2            29  
3             4  
4            11

## Blockers

Now we have the tables ready for blocking! I will introduce 4 types blockers provided by the py_entitymatching library.
1. [Attribute Equivalence Blocker](#1.-Attribute-Equivalence-Blocker)
2. [Overlap Blocker](#2.-Overlap-Blocker)
3. [Rule-based Blocker](#3.-Rule-based-Blocker)
4. [Blackbox Blocker](#4.-Blackbox-Blocker)

Each blocker can operate on ONE already-blocked-table using block_candset(), or operate on TWO separate tables using block_tables(). I'll explain their usage below.

## 1. Attribute Equivalence Blocker

As the name suggests, this blocker will block a pair if the specified attributes are NOT equivalent. In the code block below, the blocker compares the ISBN13 from tableA (a.k.a. left table) with ISBN13 from tableB (a.k.a. right table). 

For each candidate, if its ISBNs are different, blocker does not include it in the output table. You can specify which attributes to be included in the output table (l\_output\_attrs and r\_output\_attrs). Note that the output attributes will have default prefix 'ltable\_' or 'rtable\_' indicating which table they originally from and I recommend not change them. The output table will also have its primary attribute '\_id'.

In [14]:
eq = em.AttrEquivalenceBlocker()
C_ISBN = eq.block_tables(A, B, l_block_attr='ISBN13', r_block_attr='ISBN13',\
    l_output_attrs=['title', 'authors', 'publisher', 'ISBN13','publishedYear','publishedMonth','pages'],\
    r_output_attrs=['title', 'authors', 'publisher', 'ISBN13','publishedYear','publishedMonth','pages'])

In [15]:
C_ISBN.shape

(1491, 17)

In [16]:
display_attr = ['ltable_title', 'ltable_authors', 'ltable_ISBN13', 'ltable_publisher', 'rtable_title', 'rtable_authors', 'rtable_ISBN13', 'rtable_publisher']

In [17]:
C_ISBN[display_attr]

ltable_title  \
0                                    Java Generics and Collections: Speed Up the Java Development Process   
1                                                        Iron-Clad Java: Building Secure Web Applications   
2                                                          Java SE 8 for the Really Impatient / Edition 1   
3                                                          Doing Java: An Anthropological Detective Story   
4                                                            Beginning Java 2 SDK 1.4 Edition / Edition 1   
5                                                                   Sams Teach Yourself Java 6 in 21 Days   
6                                           Java 2 in Record Time: Teach Yourself the 16 Essential Skills   
7                                                                                    Big Java / Edition 3   
8                                         Active Java: Object-Oriented Programming for the World Wide Web   
9                                                                                             Art Of Java   
10                                          Teach Yourself Internet Game Programming with Java in 21 Days   
11                                                                             Professional Java Security   
12                                                                                     Java 2 For Dummies   
13                                                       Programming Mobile Objects with Java with CD-ROM   
14                                                     Platinum Edition Using HTML 3.2, Java 1.1, and CGI   
15                                                                      Java Programming with Oracle SQLJ   
16                             Inside Servlets: Server-Side Programming for the Java Platform / Edition 2   
17                                                                     Java 2 for Professional Developers   
18                                                      .NET Development for Java Programmers / Edition 1   
19                                                                                     Java and Databases   
20                                                                                     Java and Databases   
21                                  The Complete Guide to Java Database Programming (Java Masters Series)   
22                                  The Complete Guide to Java Database Programming (Java Masters Series)   
23                                                                           Better, Faster, Lighter Java   
24                                                                   Neural Network Programming with Java   
25                                                     Java Design: Objects, UML, and Process / Edition 1   
26                                                                         Java: An Object First Approach   
27                                                                                      Java Cryptography   
28                                                                     Java Transaction Design Strategies   
29                        Java Development on PDAs: Developing Applications for PocketPC and Palm Devices   
...                                                                                                   ...   
1461                                                                             Web Database Primer Plus   
1462                                                         Multimedia and Imaging Databases / Edition 1   
1463                            Database Design, Application Development, And Administration, 5th Edition   
1464                                                            Polymers: A Property Database / Edition 2   
1465                                                                              Designing XML Databases   
1466                                                        

In [18]:
C_ISBN.dtypes

_id                       int64
ltable_id                 int64
rtable_id                 int64
ltable_title             object
ltable_authors           object
ltable_publisher         object
ltable_ISBN13             int64
ltable_publishedYear      int64
ltable_publishedMonth     int64
ltable_pages              int64
rtable_title             object
rtable_authors           object
rtable_publisher         object
rtable_ISBN13             int64
rtable_publishedYear      int64
rtable_publishedMonth     int64
rtable_pages              int64
dtype: object

Look at above to get an idea of how the output table looks like. I only display 8 of 17 attributes so you don't have to scroll all the way to the right to compare the result.

This blocker is useful if each table has one attribute that is universally unique (student ID, SSN etc.). We are actually 'cheating' here to use the ISBN to reduce the candidate pairs from 12 million to 1491. In real-world datasets, you are not likely to have a universally unique attribute for your data sets.

## 2. Overlap Blocker

Overlap blocker also operates on two attributes, one from the left table and one from the right table. It will first split the attribute into string tokens by punctuations. Then it will ignore stop words (recall HW3 text classification) and count the overlapped words between the two attributes (word_level=True, overlap_size=3). 

For all candidate pairs, if the number of overlapped words is smaller than the threshold, blocker does not include them in the output table. In my case, the titles must have overlapped words greater or equal to 3.

In [19]:
# First way: using overlap blocker
ob = em.OverlapBlocker()
# block using name
C_title = ob.block_tables(A, B, 'title', 'title', word_level=True, overlap_size=3,\
    l_output_attrs=['title', 'authors', 'ISBN13','pages', 'publisher', 'publishedYear', 'publishedMonth'],\
    r_output_attrs=['title', 'authors', 'ISBN13','pages', 'publisher', 'publishedYear', 'publishedMonth'])

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:04


In [20]:
C_title.shape

(258718, 17)

In [21]:
C_title[display_attr]

ltable_title  \
0                                     The Complete Guide to Java Database Programming (Java Masters Series)   
1       A Selection of Books on the East: In the Same Series and Similarly Illustrated; British North Bo...   
2                                                           The J2EE Tutorial (The Java Series) / Edition 2   
3       An American Merchant in Europe, Asia, and Australia; a Series of Letters from Java, Singapore, C...   
4       Designing Web Services with the J2EE 1.4 Platform (The Java Series): JAX-RPC, SOAP, and XML Tech...   
5       An American Merchant in Europe, Asia and Australia: A Series of Letters From Java, Singapore, Ch...   
6                                                    Java: The Complete Reference, J2se (References Series)   
7       J2EE Technology in Practice: Building Business Applications with the Java 2 Platform (Addison-We...   
8                                     The Complete Guide to Java Database Programming (Java Masters Series)   
9                                                 The Java Virtual Machine Specification, Java SE 7 Edition   
10                                    The Java Virtual Machine Specification, Java SE 8 Edition / Edition 1   
11                                           The Java Language Specification, Java SE 8 Edition / Edition 1   
12                                           The Java Language Specification, Java SE 7 Edition / Edition 4   
13                                                              The Java Language Specification / Edition 2   
14                                                                   The Java Virtual Machine Specification   
15                                                                     The Real-time Specification for Java   
16                                                              The Java Language Specification / Edition 3   
17      The Java (programming language) Handbook - Everything You Need To Know About Java (programming l...   
18      A Dictionary of the Malay Tongue, as Spoken in the Peninsula of Malacca, the Islands of Sumatra,...   
19                                                                The Java Programming Language / Edition 4   
20                                                                The Java Programming Language / Edition 3   
21                      The JR Programming Language: Concurrent Programming in an Extended Java / Edition 1   
22                                                               A Dictionary Of The Sunda Language Of Java   
23                                            Sun Performance and Tuning: Java and the Internet / Edition 2   
24                                                            Java EE 7 Performance Tuning and Optimization   
25                                    Optimizing Java: Practical Techniques for Improved Performance Tuning   
26           Java Software Solutions : Foundations of Program Design, Update, JavaPlace Edition / Edition 2   
27                                       Java Software Solutions: Foundations of Program Design / Edition 2   
28      MyProgrammingLab with Pearson eText -- Access Code Card -- for Java Software Solutions: Foundati...   
29                         Java Foundations: Introduction to Program Design and Data Structures / Edition 4   
...                                                                                                     ...   
258688  Java Data Mining: Strategy, Standard, and Practice: A Practical Guide for Architecture, Design, ...   
258689                                   SQL and Relational Basics: A Practical Guide for PC Database Users   
258690  Spatial Database for GPS Wildlife Tracking Data: A Practical Guide to Creating a Data Management...   
258691                       Principles of Database Systems with Internet and Java Applications / Edition 1   
258692                                                          Principles and Practice of Database

As you can see, the output table is really large (250K rows), but we've successfully reduced the size of candidate pairs from 12 million to 250K. We can further reduce the size by chaining other blockers to our result C_title table.

## 3. Rule-based Blocker

Before we chain the result from overlap blocker. Let me introduce the general usage of rule-based blocker first.

The py_entitymatching library provides a feature table for measuring the similarity between attributes from two tables. See below:

In [22]:
feature_table = em.get_features_for_blocking(A, B, validate_inferred_attr_types=False) # set to True for development

In [23]:
feature_table

feature_name  left_attribute right_attribute  \
0                                 id_id_exm              id              id   
1                                 id_id_anm              id              id   
2                            id_id_lev_dist              id              id   
3                             id_id_lev_sim              id              id   
4               title_title_jac_qgm_3_qgm_3           title           title   
5           title_title_cos_dlm_dc0_dlm_dc0           title           title   
6                           title_title_mel           title           title   
7                      title_title_lev_dist           title           title   
8                       title_title_lev_sim           title           title   
9           authors_authors_jac_qgm_3_qgm_3         authors         authors   
10      authors_authors_cos_dlm_dc0_dlm_dc0         authors         authors   
11      authors_authors_jac_dlm_dc0_dlm_dc0         authors         authors   
12                      authors_authors_mel         authors         authors   
13                 authors_authors_lev_dist         authors         authors   
14                  authors_authors_lev_sim         authors         authors   
15                      authors_authors_nmw         authors         authors   
16                       authors_authors_sw         authors         authors   
17                        ISBN13_ISBN13_exm          ISBN13          ISBN13   
18                        ISBN13_ISBN13_anm          ISBN13          ISBN13   
19                   ISBN13_ISBN13_lev_dist          ISBN13          ISBN13   
20                    ISBN13_ISBN13_lev_sim          ISBN13          ISBN13   
21                          pages_pages_exm           pages           pages   
22                          pages_pages_anm           pages           pages   
23                     pages_pages_lev_dist           pages           pages   
24                      pages_pages_lev_sim           pages           pages   
25      publisher_publisher_jac_qgm_3_qgm_3       publisher       publisher   
26  publisher_publisher_cos_dlm_dc0_dlm_dc0       publisher       publisher   
27  publisher_publisher_jac_dlm_dc0_dlm_dc0       publisher       publisher   
28                  publisher_publisher_mel       publisher       publisher   
29             publisher_publisher_lev_dist       publisher       publisher   
30              publisher_publisher_lev_sim       publisher       publisher   
31                  publisher_publisher_nmw       publisher       publisher   
32                   publisher_publisher_sw       publisher       publisher   
33          publishedYear_publishedYear_exm   publishedYear   publishedYear   
34          publishedYear_publishedYear_anm   publishedYear   publishedYear   
35     publishedYear_publishedYear_lev_dist   publishedYear   publishedYear   
36      publishedYear_publishedYear_lev_sim   publishedYear   publishedYear   
37        publishedMonth_publishedMonth_exm  publishedMonth  publishedMonth   
38        publishedMonth_publishedMonth_anm  publishedMonth  publishedMonth   
39   publishedMonth_publishedMonth_lev_dist  publishedMonth  publishedMonth   
40    publishedMonth_publishedMonth_lev_sim  publishedMonth  publishedMonth   
41            publishedDay_publishedDay_exm    publishedDay    publishedDay   
42            publishedDay_publishedDay_anm    publishedDay    publishedDay   
43       publishedDay_publishedDay_lev_dist    publishedDay    publishedDay   
44        publishedDay_publishedDay_lev_sim    publishedDay    publishedDay   

   left_attr_tokenizer right_attr_tokenizer       simfunction  \
0                 None                 None       exact_match   
1                 None                 None          abs_norm   
2                 None                 None          lev_dist   
3                 None                 None           lev_sim   
4                qgm_3                qgm_3           jaccard   
5     

We can add these rules to our rule-based blockers. Here I introduce the feature 'authors_authors_jac_qgm_3_qgm_3'. authors_authors means I'm comparing authors from left table to authors from the right table. 'jac' means Jaccard index or **Intersection over Union** (https://en.wikipedia.org/wiki/Jaccard_index). 'qgm_3' means we are using n-gram model (3 characters level instead of word-level we've seen). 

To combine these concepts, I'll show you a simple example. Suppose we have A = dave, B = dav, using 3 gram character level with padding # at front and back, we have:

A_set: {##d, #da, dav, ave, ve#, e##}, 

B_set: {##d, #da, dav, av#, v##}. 

intersection: {##d, #da, dav}

union: {##d, #da, dav, ave, ve#, e##, av#, v##}. Thus the Jaccard index is 3/8.

Here I build a rule blocker and add a rule "block pairs that have jaccard index of authors smaller than 0.4". You can add as many rules as you wish. 

You can name your rule using 'rule_name=' when you add it, then later on you can delete this rule by calling delete_rule(rule name). 'n_jobs=' is also useful as it specifies the number of parallel jobs (-1 means use all your CPUs).

In [24]:
rule_blocker=em.RuleBasedBlocker() 
rule_blocker.add_rule(['authors_authors_jac_qgm_3_qgm_3(ltuple, rtuple) < 0.4'], feature_table, rule_name='author_r')
C_author_title = rule_blocker.block_candset(C_title, n_jobs=-1)

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:08


In [25]:
C_author_title.shape

(3246, 17)

In [26]:
C_author_title[display_attr]

ltable_title  \
11                                           The Java Language Specification, Java SE 8 Edition / Edition 1   
12                                           The Java Language Specification, Java SE 7 Edition / Edition 4   
13                                                              The Java Language Specification / Edition 2   
16                                                              The Java Language Specification / Edition 3   
26           Java Software Solutions : Foundations of Program Design, Update, JavaPlace Edition / Edition 2   
27                                       Java Software Solutions: Foundations of Program Design / Edition 2   
28      MyProgrammingLab with Pearson eText -- Access Code Card -- for Java Software Solutions: Foundati...   
30                      Java Software Solutions (Java 5 version): Foundations of Program Design / Edition 4   
31            Java Software Solutions: Foundations of Program Design, CodeMate Enhanced Edition / Edition 3   
32                                       Java Software Solutions: Foundations of Program Design / Edition 7   
36                                       Java Software Solutions: Foundations of Program Design / Edition 5   
37                                       Java Software Solutions: Foundations of Program Design / Edition 6   
40                         Java Software Solutions with MyProgrammingLab -- Access Card Package / Edition 8   
41                                                                      Java Software Solutions / Edition 8   
42                                            Student Value Edition for Java Software Solutions / Edition 8   
43                                            Java Software Solutions AP and AWS Review Package / Edition 1   
73                                                   Data Structures and Abstractions with Java / Edition 1   
129                                                  Data Structures and Abstractions with Java / Edition 2   
246                                                                                           Java and XSLT   
250          Student Value Edition for Java: An Introduction to Problem Solving and Programming / Edition 6   
292     Java: An Introduction to Problem Solving and Programming with MyProgrammingLab -- Access Card Pa...   
304                                    Java: An Introduction to Problem Solving and Programming / Edition 7   
329     Java: An Introduction to Computer Science and Programming JDK 1.2 Compatible Source Code / Editi...   
404          Student Value Edition for Java: An Introduction to Problem Solving and Programming / Edition 7   
589                                                            Core Java Volume I--Fundamentals / Edition 9   
591                                                     Core Java, Volume II--Advanced Features / Edition 9   
593                                                           Core Java Volume I--Fundamentals / Edition 10   
594                                                    Core Java, Volume II--Advanced Features / Edition 10   
596                                                                   Core Java 2: Fundamentals / Edition 7   
600                                                Core Java 2: Volume II, Advanced Features, Fifth Edition   
...                                                                                                     ...   
257792                                                                           SSAT Database and 11 Tests   
257796                                                                  Database Development in Lotus Notes   
257807                                                      World Databases in Biosciences and Pharmacology   
257809                                                               World Databases in Physics-Mathematics   
257810                                                                   World Databases in Social 

Below shows how to delete rule, add rule and display all rules.

In [27]:
rule_blocker.delete_rule('author_r')

True

In [28]:
rule_blocker.add_rule(['authors_authors_jac_qgm_3_qgm_3(ltuple, rtuple) < 0.4'], feature_table, rule_name='author_r')
rule_blocker.add_rule(['publisher_publisher_jac_qgm_3_qgm_3(ltuple, rtuple) < 0.4'], feature_table, rule_name='publisher_r')
rule_blocker.get_rule_names()

odict_keys(['author_r', 'publisher_r'])

Here, instead of blocking on tableA and tableB, I use the already-blocked table C_title from the last blocker to further reduce the size. I also added publisher jaccard index in the rules.

Scroll down to see what the output looks like.

In [29]:
# rule-based block on merged data
C_publisher_author_title = rule_blocker.block_candset(C_title, n_jobs=-1)

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:08


In [30]:
C_publisher_author_title.shape

(1447, 17)

In [31]:
C_publisher_author_title[display_attr]

ltable_title  \
11                                           The Java Language Specification, Java SE 8 Edition / Edition 1   
12                                           The Java Language Specification, Java SE 7 Edition / Edition 4   
13                                                              The Java Language Specification / Edition 2   
16                                                              The Java Language Specification / Edition 3   
36                                       Java Software Solutions: Foundations of Program Design / Edition 5   
37                                       Java Software Solutions: Foundations of Program Design / Edition 6   
73                                                   Data Structures and Abstractions with Java / Edition 1   
129                                                  Data Structures and Abstractions with Java / Edition 2   
246                                                                                           Java and XSLT   
589                                                            Core Java Volume I--Fundamentals / Edition 9   
591                                                     Core Java, Volume II--Advanced Features / Edition 9   
593                                                           Core Java Volume I--Fundamentals / Edition 10   
594                                                    Core Java, Volume II--Advanced Features / Edition 10   
596                                                                   Core Java 2: Fundamentals / Edition 7   
604                                  Big Java: Early Objects, Binder Ready Version, 6th Edition / Edition 6   
605                                                             Java For Everyone: Late Objects / Edition 2   
607                                                                      Big Java: Late Objects / Edition 1   
608                                                                     Big Java: Early Objects / Edition 5   
632                                                                       Just Java 2: J2SE 1.5 / Edition 6   
682                                                                   Learn Objective-C for Java Developers   
767                                           Sun Performance and Tuning: Java and the Internet / Edition 2   
930                                    Configuring and Tuning Databases on the Solaris Platform / Edition 1   
931                                                          Java SE 8 for the Really Impatient / Edition 1   
955                                                                 Core Java for the Impatient / Edition 1   
1001                                                                  Sams Teach Yourself Java 6 in 21 Days   
1006                                      Sams Teach Yourself Java in 21 Days (Covering Java 7 and Android)   
1008    Java in 24 Hours, Sams Teach Yourself (Covering Java 8) Barnes & Noble Exclusive Edition / Editi...   
1009                                                Java in 24 Hours, Sams Teach Yourself (Covering Java 8)   
1010                                     Java in 21 Days, Sams Teach Yourself (Covering Java 8) / Edition 7   
1011                                     Sams Teach Yourself Java 1.2 in 21 Days: Complete Compiler Edition   
...                                                                                                     ...   
256098                                                     Business and Management: Bibliographic Databases   
256163                                                    Bioinformatics: Databases and Systems / Edition 1   
256169                                                                Bioinformatics: Databases and Systems   
256451                                                 Natural Medicines Comprehensive Database / Edition 3   
256453                                                             Natural Medicines Comprehensive 

The result is quite satisfying, we reduce the 250K C_title to 1447 rows and you can see the candidates looks similar.

## Bonus: Rule-based Blocker 

Below is the standard usage of rule-based blockers on two tables. The interfaces is nearly the same to other blockers.

In [32]:
# rule-based on A,B sets
C_publisher_author = rule_blocker.block_tables(A, B, n_jobs=-1,\
    l_output_attrs=['title', 'authors', 'ISBN13','pages', 'publisher', 'publishedYear', 'publishedMonth'],\
    r_output_attrs=['title', 'authors', 'ISBN13','pages', 'publisher', 'publishedYear', 'publishedMonth'])

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:00


Finding pairs with missing value...


0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:00:00


In [33]:
C_publisher_author[display_attr]

ltable_title  \
0                                         The Java Language Specification, Java SE 8 Edition / Edition 1   
1                                                            The Java Language Specification / Edition 3   
2                                         The Java Language Specification, Java SE 7 Edition / Edition 4   
3                                                            The Java Language Specification / Edition 2   
8                                     Java Software Solutions: Foundations of Program Design / Edition 5   
15                                    Java Software Solutions: Foundations of Program Design / Edition 6   
16                                                Data Structures and Abstractions with Java / Edition 1   
17                                                Data Structures and Abstractions with Java / Edition 2   
20                                     Java: Introduction to Problem Solving and Programming / Edition 5   
21                                                                     Java Extreme Programming Cookbook   
22                                                                                         Java and XSLT   
31                                                          Core Java Volume I--Fundamentals / Edition 9   
32                                                                 Core Java 2: Fundamentals / Edition 7   
37                                                   Core Java, Volume II--Advanced Features / Edition 9   
43                                                         Core Java Volume I--Fundamentals / Edition 10   
45                                                  Core Java, Volume II--Advanced Features / Edition 10   
57                                                                                  Big Java / Edition 3   
60                                                              Java Concepts: Early Objects / Edition 7   
61                                Big Java: Early Objects, Binder Ready Version, 6th Edition / Edition 6   
62                                        Java Concepts, Binder Ready Version: Early Objects / Edition 8   
64                                                           Java For Everyone: Late Objects / Edition 2   
66                                                                    Big Java: Late Objects / Edition 1   
68                                                                   Big Java: Early Objects / Edition 5   
71                                                   Computing Concepts with Java Essentials / Edition 3   
72                                                     Java Concepts for AP Computer Science / Edition 5   
73                                                 Computing Concepts with Java 2 Essentials / Edition 2   
74                                                   Computing Concepts with Java Essentials / Edition 1   
75                                            Java Concepts: Compatible with Java 5, 6 and 7 / Edition 6   
76                            Java Concepts: Advanced Placement Computer Science Study Guide / Edition 5   
82                                                                         Java For Everyone / Edition 1   
..                                                                                                   ...   
500                                                    Recovery in Parallel Database Systems / Edition 2   
531                                                           Databases and Mobile Computing / Edition 1   
533                                                          MYSQL/PHP Database Applications / Edition 2   
534                                                                      MySQL/PHP Database Applications   
535                                                                          Datalog and Logic Databases   
536                                                    Database Marketing im Finanzdienstleistun

## 4. Blackbox Blocker

This is the last blocker and the most interesting one in my opinion. You have the complete freedom of customizing the rule of blocking!

In [34]:
# Define the black_box function used to check the equivalence between numerical attributes
def published_range_block(x, y):
    month_diff = abs(x['publishedMonth'] - y['publishedMonth'])
    year_diff = abs(x['publishedYear'] - y['publishedYear'])
    pages_diff = abs(x['pages'] - y['pages'])
    if month_diff > 2 or year_diff > 2 or pages_diff > 10:
        return True
    else:
        return False

You can design a function for blocking. **For conditions that you think should be blocked, return True, otherwise return False**. In the above code, I set the rules as:
1. "if publishedMonth difference is greater than 2, block"
2. "if publishedYear difference is greater than 2, block"
3. "if pages difference is greater than 10, block"

In [35]:
bb1 = em.BlackBoxBlocker()
bb1.set_black_box_function(published_range_block)
C_pub_date = bb1.block_tables(A, B, n_jobs=-1,\
    l_output_attrs=['title', 'authors', 'publisher', 'ISBN13','publishedYear','publishedMonth','pages'],\
    r_output_attrs=['title', 'authors', 'publisher', 'ISBN13','publishedYear','publishedMonth','pages'])

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:02:49


In [36]:
# Checking the output of the date blocker
C_pub_date.shape

(18813, 17)

In [37]:
C_pub_date

_id  ltable_id  rtable_id  \
0          0          2        107   
1          1          2        599   
2          2          3        320   
3          3          3        542   
4          4          4        306   
5          5          4        312   
6          6          4        719   
7          7          5        271   
8          8          5        441   
9          9          5        562   
10        10          6        115   
11        11          7        211   
12        12          7        500   
13        13         10          9   
14        14         10        187   
15        15         10        607   
16        16         14        283   
17        17         14        534   
18        18         16        317   
19        19         16        334   
20        20         16        560   
21        21         16        829   
22        22         17        450   
23        23         19        482   
24        24         21        174   
25        25         21        229   
26        26         22        162   
27        27         23         93   
28        28         23        227   
29        29         23        319   
...      ...        ...        ...   
18783  18783       5245       3407   
18784  18784       5246       3683   
18785  18785       5247       2857   
18786  18786       5247       2955   
18787  18787       5247       3013   
18788  18788       5247       3106   
18789  18789       5247       3131   
18790  18790       5251       3450   
18791  18791       5259       3263   
18792  18792       5259       3573   
18793  18793       5259       3597   
18794  18794       5261       3769   
18795  18795       5262       2914   
18796  18796       5262       2934   
18797  18797       5266       2972   
18798  18798       5266       3059   
18799  18799       5267       3295   
18800  18800       5271       3450   
18801  18801       5272       2880   
18802  18802       5272       3155   
18803  18803       5272       3260   
18804  18804       5272       3366   
18805  18805       5272       3660   
18806  18806       5272       3667   
18807  18807       5274       3243   
18808  18808       5274       3312   
18809  18809       5274       3423   
18810  18810       5274       3505   
18811  18811       5274       3506   
18812  18812       5274       3650   

                                                                                              ltable_title  \
0                                      Java Programs to Accompany Programming Logic and Design / Edition 8   
1                                      Java Programs to Accompany Programming Logic and Design / Edition 8   
2                                         Raspberry Pi with Java: Programming the Internet of Things (IoT)   
3                                         Raspberry Pi with Java: Programming the Internet of Things (IoT)   
4                                     Java Generics and Collections: Speed Up the Java Development Process   
5                                     Java Generics and Collections: Speed Up the Java Development Process   
6                                     Java Generics and Collections: Speed Up the Java Development Process   
7                                                         Iron-Clad Java: Building Secure Web Applications   
8                                                         Iron-Clad Java: Building Secure Web Applications   
9                                                         Iron-Clad Java: Building Secure Web Applications   
10                                                          Java SE 8 for the Really Impatient / Edition 1   
11                           Programming Groovy 2: Dynamic Productivity for the Java Developer / Edition 1   
12                           Programming Groovy 2: Dynamic Productivity for the Java Developer / Edition 1   
13                                                Algorithms in Java, Part 5: Graph Algorithms / Edition 3   
14

Another example function is to block on partial ISBN as ISBN heads indicate publishers (assume you don't want to cheat on comparing entire ISBN).

In [38]:
def partial_isbn_block(x,y):
    isbn_A = str((x['ISBN13']))
    isbn_head_A = isbn_A[0:6]
    isbn_B = str(y['ISBN13'])
    isbn_head_B = isbn_B[0:6]
    # compare first 6 digits
    if isbn_head_A != isbn_head_B:
        return True
    else:
        return False 

## Blackbox Blocker: things to remember

1. Although blackbox blocker is great for customization, it is computational intensive! It takes **minutes** to finish blocking while other blockers finish in seconds on my data set.
2. Blackbox blocker can only accept one function at a time, meaning you cannot add many functions like the rule-based blocker. 

What I recommend here is to write several blocker functions and merge them into a combined blocker function using 'or' to combine conditions. Below is the function that I combined the published range function and partial ISBN function.

In [39]:
def combined_block(x, y):
    month_diff = abs(x['publishedMonth'] - y['publishedMonth'])
    year_diff = abs(x['publishedYear'] - y['publishedYear'])
    pages_diff = abs(x['pages'] - y['pages'])
    isbn_A = str((x['ISBN13']))
    isbn_head_A = isbn_A[0:6]
    isbn_B = str(y['ISBN13'])
    isbn_head_B = isbn_B[0:6]
    
    if month_diff > 2 or year_diff > 2 or pages_diff > 10 or isbn_head_A != isbn_head_B:
        return True
    else:
        return False

In [40]:
bb2 = em.BlackBoxBlocker()
# one function only
bb2.set_black_box_function(combined_block)
C_combined = bb2.block_tables(A, B, n_jobs=-1,\
    l_output_attrs=['title', 'authors', 'publisher', 'ISBN13','publishedYear','publishedMonth','pages'],\
    r_output_attrs=['title', 'authors', 'publisher', 'ISBN13','publishedYear','publishedMonth','pages'])

0% [##############################] 100% | ETA: 00:00:00
Total time elapsed: 00:03:24


In [41]:
C_combined.shape

(1746, 17)

In [42]:
C_combined[display_attr]

ltable_title  \
0                                    Java Generics and Collections: Speed Up the Java Development Process   
1                                                        Iron-Clad Java: Building Secure Web Applications   
2                                               Java Gently: Programming Principles Explained / Edition 3   
3                                                          Doing Java: An Anthropological Detective Story   
4                                                            Beginning Java 2 SDK 1.4 Edition / Edition 1   
5                                                                   Sams Teach Yourself Java 6 in 21 Days   
6                                         Active Java: Object-Oriented Programming for the World Wide Web   
7                                           Teach Yourself Internet Game Programming with Java in 21 Days   
8                 Women of the Kakawin World: Marriage and Sexuality in the Indic Courts of Java and Bali   
9                 Women of the Kakawin World: Marriage and Sexuality in the Indic Courts of Java and Bali   
10                                                                                     Java 2 For Dummies   
11                            Java Performance for Windows: Using Microsoft AFC, NFC, and XML with CD-ROM   
12                                                                      Java Programming with Oracle SQLJ   
13                                                                                     Java and Databases   
14                                                                           Better, Faster, Lighter Java   
15                                                                           Better, Faster, Lighter Java   
16                                                                                      Java Cryptography   
17                                                                                      Java Cryptography   
18                                                                     Java Transaction Design Strategies   
19                                            Enterprise Java Security: Building Secure J2EE Applications   
20                                   Java the UML Way: Integrating Object-Oriented Design and Programming   
21                                               Data Structures and Other Objects Using Java / Edition 3   
22                                                                   C++ For Java Programmers / Edition 1   
23                       Ready-Made Java 2 Applications for File Maintenance: Templates 2000 for Java 1.2   
24                                                    Gentry: Social Change in Java: the Tale of a Family   
25                                     Java Software Solutions: Foundations of Program Design / Edition 2   
26    The JavaScript Developer's Resource: Client-Side Programming Using HTML, Netscape Plug-Ins and J...   
27    The JavaScript Developer's Resource: Client-Side Programming Using HTML, Netscape Plug-Ins and J...   
28                                               Concurrent and Distributed Computing in Java / Edition 1   
29                                                                            Exploring Java, 2nd Edition   
...                                                                                                   ...   
1716                                                                                    Spatial Databases   
1717                                                                                Intelligent Databases   
1718                  Database Systems: Design, Implementation, and Management, Fifth Edition / Edition 5   
1719                                                        High-Performance Oracle Database Applications   
1720                 Client/Server Database Design with SYBASE: A High-Performance and Fine- Tuning Guide   
1721                                                   Black

Finally, when you have different output tables and you want to merge them, call the below function, it will union the tables.

In [43]:
C = em.combine_blocker_outputs_via_union([C_publisher_author_title, C_combined])

In [44]:
C.shape

(2765, 17)

In [45]:
C

_id  ltable_id  rtable_id  \
0        0          0        443   
1        1          4        312   
2        2          5        441   
3        3          6         15   
4        4          6        273   
5        5          7       2980   
6        6         10        219   
7        7         10        501   
8        8         16       2359   
9        9         17        450   
10      10         17       2896   
11      11         19        482   
12      12         21        229   
13      13         23         16   
14      14         23         93   
15      15         23        132   
16      16         23        365   
17      17         23        379   
18      18         23        473   
19      19         23       1097   
20      20         25       1313   
21      21         27        342   
22      22         27       1585   
23      23         31        210   
24      24         32        419   
25      25         32       1906   
26      26         34         34   
27      27         34        356   
28      28         35        329   
29      29         36       2483   
...    ...        ...        ...   
2735  2735       5179       2979   
2736  2736       5188        774   
2737  2737       5188       1645   
2738  2738       5188       3095   
2739  2739       5190       2518   
2740  2740       5191       1992   
2741  2741       5195       1413   
2742  2742       5197       3720   
2743  2743       5202       2783   
2744  2744       5202       3461   
2745  2745       5208       2651   
2746  2746       5218       3421   
2747  2747       5230       3396   
2748  2748       5234       2854   
2749  2749       5237       1141   
2750  2750       5239       2300   
2751  2751       5241       3476   
2752  2752       5242       2906   
2753  2753       5243       2147   
2754  2754       5250       2246   
2755  2755       5251       1720   
2756  2756       5252       2354   
2757  2757       5252       2481   
2758  2758       5252       3066   
2759  2759       5253       2831   
2760  2760       5262       2303   
2761  2761       5268       1638   
2762  2762       5271       2033   
2763  2763       5271       2116   
2764  2764       5274       3505   

                                                                                             ltable_title  \
0                                             The Java Tutorial: A Short Course on the Basics / Edition 6   
1                                    Java Generics and Collections: Speed Up the Java Development Process   
2                                                        Iron-Clad Java: Building Secure Web Applications   
3                                                          Java SE 8 for the Really Impatient / Edition 1   
4                                                          Java SE 8 for the Really Impatient / Edition 1   
5                           Programming Groovy 2: Dynamic Productivity for the Java Developer / Edition 1   
6                                                Algorithms in Java, Part 5: Graph Algorithms / Edition 3   
7                                                Algorithms in Java, Part 5: Graph Algorithms / Edition 3   
8     Cameron Mckenzie's SCJA Sun Certified Java Associate: Certification Study Guide for Jave 5, J2EE...   
9                                               Java Gently: Programming Principles Explained / Edition 3   
10                                              Java Gently: Programming Principles Explained / Edition 3   
11                                                         Doing Java: An Anthropological Detective Story   
12                                                           Beginning Java 2 SDK 1.4 Edition / Edition 1   
13                                                                  Sams Teach Yourself Java 6 in 21 Days   
14                                                                  Sams Teach Yourself Java 6 in 21 Days   
15                            

## What's Next

I'll introduce you how to modify the output to get ready for applying supervised machine learning techniques on data matching.

The code below random sampled 400 rows from the tableC (the table we unioned in the last step).

In [46]:
S = em.sample_table(C, 400)

Below function extends the sample with an extra column filled with 0 called 'label', which can be used to indicate whether the left table and right table matches (to be done manually). 

In [47]:
S = em.label_table(S, 'label')

Column name (label) is not present in dataframe


In [48]:
S

_id  ltable_id  rtable_id  \
4        4          6        273   
13      13         23         16   
46      46         78        388   
50      50         90        567   
64      64        116       1384   
67      67        119        684   
71      71        126       1751   
80      80        138        631   
84      84        142         83   
86      86        143        784   
87      87        146         85   
92      92        151        598   
117    117        198        481   
121    121        198       1324   
123    123        203       1236   
126    126        206          8   
129    129        206       1675   
133    133        208        115   
134    134        210        205   
136    136        211       1891   
141    141        212       1221   
147    147        222        453   
148    148        223        600   
152    152        235        828   
153    153        235       2276   
158    158        238       2102   
160    160        239       2814   
165    165        253       2980   
170    170        261       1823   
175    175        277        558   
...    ...        ...        ...   
2549  2549       4600       2194   
2551  2551       4600       2377   
2559  2559       4622       2265   
2561  2561       4623       2371   
2573  2573       4661        749   
2585  2585       4694       2440   
2586  2586       4694       2525   
2592  2592       4716        586   
2596  2596       4725       1981   
2608  2608       4759       2359   
2630  2630       4812       1902   
2632  2632       4813       2945   
2637  2637       4825       2481   
2640  2640       4831       3540   
2641  2641       4844        159   
2660  2660       4883       2786   
2664  2664       4889       1970   
2670  2670       4929       1955   
2674  2674       4933       3211   
2679  2679       4944       2671   
2684  2684       4954       3694   
2693  2693       4983       2342   
2698  2698       4997       2366   
2713  2713       5067        362   
2722  2722       5109       3592   
2726  2726       5150       2054   
2727  2727       5152       3664   
2735  2735       5179       2979   
2736  2736       5188        774   
2744  2744       5202       3461   

                                                                                             ltable_title  \
4                                                          Java SE 8 for the Really Impatient / Edition 1   
13                                                                  Sams Teach Yourself Java 6 in 21 Days   
46                                                                           Better, Faster, Lighter Java   
50                                                                     Java Transaction Design Strategies   
64                                                                     Programming the Internet with Java   
67                       Ready-Made Java 2 Applications for File Maintenance: Templates 2000 for Java 1.2   
71                                            Interdisciplinary Computing in Java Programming / Edition 1   
80    The JavaScript Developer's Resource: Client-Side Programming Using HTML, Netscape Plug-Ins and J...   
84                                               Concurrent and Distributed Computing in Java / Edition 1   
86                                                                            Exploring Java, 2nd Edition   
87                                                          Pro JPA 2: Mastering the Java Persistence API   
92                                                               Java Network Programming, Second Edition   
117                                                                 Jasperreports 3.5 For Java Developers   
121                                                                 Jasperreports 3.5 For Java Developers   
123                                                  Guide to Java: A Concise Introduction to Programming   
126                           

In [49]:
# save new tables to csv file
em.to_csv_metadata(C, './tableC.csv')
em.to_csv_metadata(S, './tableS.csv')

True

Below demonstrates how you can split the labeled set into training test and test set. I didn't label so the label column is all zero. 'train_proportion=' specifies the portion you want to put into the training set, the rest will be put into the test set.

In [50]:
# assume tableS is labeled now
path_S = os.path.join(cur_path, 'tableS.csv')
labeled_S = em.read_csv_metadata(path_S, ltable=A, rtable=B)
sets = em.split_train_test(labeled_S, train_proportion=0.4, random_state=0)

In [51]:
sets['train']

_id  ltable_id  rtable_id  \
397  2735       5179       2979   
118   931       1375       2772   
254  1776       2821       3217   
188  1346       2041        708   
208  1460       2296       2505   
375  2585       4694       2440   
110   826       1233       1406   
149  1147       1726        138   
157  1200       1798        692   
152  1176       1759        125   
16    129        206       1675   
269  1870       3012       3524   
75    571        805       1623   
109   814       1200       1838   
327  2220       3789        799   
205  1431       2203       3432   
315  2125       3553        221   
139  1088       1626        160   
237  1667       2600       1305   
319  2142       3584       2260   
248  1747       2770       3603   
308  2114       3523       2843   
19    136        211       1891   
226  1586       2510       3591   
306  2093       3486       2810   
3      50         90        567   
276  1928       3158       1968   
125   982       1461       1871   
77    594        853        772   
184  1335       2026       1476   
..    ...        ...        ...   
243  1691       2634       3366   
115   917       1352        736   
265  1848       2971        346   
72    538        766       1095   
333  2251       3840       3691   
25    158        238       2102   
165  1245       1869       2120   
337  2270       3898       2494   
384  2641       4844        159   
174  1291       1950        127   
386  2664       4889       1970   
39    229        355        333   
193  1380       2081       2112   
314  2124       3552       2730   
88    681        979        528   
70    530        760       1622   
87    679        976        735   
292  2035       3364       3177   
242  1680       2615       1639   
277  1952       3215       2808   
211  1476       2330        619   
9      86        143        784   
359  2448       4337       3085   
195  1383       2083        921   
251  1754       2774       3070   
323  2173       3641       1102   
192  1377       2078       1562   
117   928       1374       1635   
47    310        443        546   
172  1278       1933       2021   

                                                                                            ltable_title  \
397                                                            The Essential Guide to Database Marketing   
118                        Java Ee6 Cookbook for Securing, Tuning, and Extending Enterprise Applications   
254                                         Oracle Database & Data Warehouse 11g: Performance Assessment   
188                                                                   Java Fundamental Classes Reference   
208  The Complete Database Marketer: Second Generation Strategies and Techniques for Tapping the Powe...   
375                                                                         Access Control For Databases   
110                                                                 Tricks of the Java Programming Gurus   
149                                                                            Rails for Java Developers   
157      Fundamentals of Java: AP* Computer Science Essentials for the A Exam, Third Edition / Edition 3   
152                       Object-Oriented Software Engineering: Using UML, Patterns and Java / Edition 2   
16                                                                     Just Java 2: J2SE 1.5 / Edition 6   
269                                                              Oracle Database 11g: SQL Fundamentals I   
75                  Pro JavaFX Platform: Script, Desktop and Mobile RIA with Java Technology / Edition 1   
109                                           Java Concepts: Compatible with Java 5, 6 and 7 / Edition 6   
327                                                              Java 2 Database Programming for Dummies   
205  Children's Books in Children's Hands: An Introduction to Their Literature (with Children's Liter...   

In [52]:
sets['test']

_id  ltable_id  rtable_id  \
132  1015       1493       1015   
309  2116       3529       3573   
341  2311       3998       2986   
196  1384       2083       2153   
246  1723       2713       2774   
60    455        642         93   
155  1197       1788       1808   
261  1832       2947       2738   
141  1098       1640        132   
214  1500       2360       3077   
37    218        341       1166   
134  1044       1560       1239   
113   870       1304        510   
348  2363       4080       2079   
12    117        198        481   
59    449        629       1618   
293  2039       3375       2589   
140  1093       1636       1736   
206  1438       2223       2505   
199  1397       2128       1636   
176  1298       1966        131   
268  1868       3007        103   
124   980       1461       1363   
344  2331       4033       2848   
175  1293       1954       1045   
313  2123       3547       3466   
78    596        861        142   
15    126        206          8   
286  2020       3325       2545   
102   773       1145        400   
..    ...        ...        ...   
67    515        747       1363   
24    153        235       2276   
332  2247       3832       3603   
44    291        421       1071   
241  1674       2611       2855   
129   991       1475        913   
93    716       1039        387   
111   828       1234        333   
166  1254       1885       1412   
389  2679       4944       2671   
383  2640       4831       3540   
342  2321       4021       3486   
40    232        358       1684   
18    134        210        205   
284  2004       3297       3524   
79    597        867        104   
249  1748       2771       2778   
394  2722       5109       3592   
71    533        762       1363   
13    121        198       1324   
367  2532       4543       3058   
213  1496       2350       2740   
385  2660       4883       2786   
388  2674       4933       3211   
228  1599       2525       2946   
160  1218       1816       1607   
104   776       1150       1624   
161  1222       1825       1075   
83    646        928        118   
189  1347       2042       3011   

                                                                                            ltable_title  \
132                                   Java Programming with MS Visual J++ 7.0: Comprehensive / Edition 2   
309  Building the Agile Database: How to Build a Successful Application Using Agile Without Sacrifici...   
341  Descriptive models, grade-tonnage relations, and databases for the assessment of sediment-hosted...   
196                                                                           Beginning C# 5.0 Databases   
246                                      A Simplified Forest Inventory and Analysis Database: FIADB-Lite   
60                                    Sams Teach Yourself Java 1.2 in 21 Days: Complete Compiler Edition   
155                                                                                  Gingerbread 'n Java   
261                                                                    Database Management Through DBase   
141                                Sams Teach Yourself Java 2 in 21 Days, Professional Reference Edition   
214                                                         Fundamentals of Database Systems / Edition 5   
37                                                                           Java Quick Syntax Reference   
134                                          A Guide to Programming in Java: For Java SE 5 and Java SE 6   
113                                                                175 Multiple-Choice Questions in Java   
348                                                  Relational Database Writings, 1985-1989 / Edition 1   
12                                                                 Jasperreports 3.5 For Java Developers   
59                 Java by Dissection: The Essentials of Java Programming, JavaPlace Edition / Edition 1   

In [53]:
sets['train'].shape

(160, 18)

In [54]:
sets['test'].shape

(240, 18)

Now you are ready to apply different supervised learning techniques on the labeled set (eg. scikit-learn).

## Summary and references

In this tutorial, I first apply pre-processing on the data set, then 4 types of blockers are introduced. You can also start working on the data matching stage by following the last part.

py_entitymatching: https://github.com/anhaidgroup/py_entitymatching/

User manual: http://anhaidgroup.github.io/py_entitymatching/v0.1.x/singlepage.html
